lightningd/peer_control.c

```diff
@@ -1016,7 +1016,8 @@ static void json_close(struct command *cmd,
 			return;
 	}
 
-	if (!channel && peer) {
+	if (!channel) {
+		assert(peer);
```

Comment:
Yech, this points out how confusing this function is.

I think this whole thing needs to be hoist into the `if (peer)` case above.  perhaps as a separate cleanup patch?